test/helpers/envtest.go

```diff
@@ -0,0 +1,141 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package helpers
+
+import (
+	"context"
+	"os"
+	"path"
+	"path/filepath"
+	goruntime "runtime"
+	"time"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/cluster-api/util/kubeconfig"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+func init() {
+	// Chdir to project root so the paths to CRDs are stable
+	_, filename, _, _ := goruntime.Caller(0) //nolint
+	dir := path.Join(path.Dir(filename), "..")
+	err := os.Chdir(dir)
+
+	if err != nil {
+		panic(err)
+	}
+}
+
+type TestEnvironment struct {
+	*envtest.Environment
+	manager.Manager
+	client.Client
+
+	Config  *rest.Config
+	scheme  *runtime.Scheme
+	doneMgr chan struct{}
+}
+
+func newTestEnvironment() *TestEnvironment {
+	scheme := scheme.Scheme
+	utilruntime.Must(apiextensionsv1.AddToScheme(scheme))
+	utilruntime.Must(clusterv1.AddToScheme(scheme))
+	utilruntime.Must(bootstrapv1.AddToScheme(scheme))
+
+	return &TestEnvironment{
+		Environment: &envtest.Environment{
+			ErrorIfCRDPathMissing: true,
+			CRDDirectoryPaths: []string{
+				filepath.Join("..", "config", "crd", "bases"),
+				filepath.Join("..", "controlplane", "kubeadm", "config", "crd", "bases"),
+				filepath.Join("..", "bootstrap", "kubeadm", "config", "crd", "bases"),
+			},
+			CRDs: []runtime.Object{
+				external.TestGenericBootstrapCRD,
+				external.TestGenericBootstrapTemplateCRD,
+				external.TestGenericInfrastructureCRD,
+				external.TestGenericInfrastructureTemplateCRD,
+			},
+		},
+		scheme: scheme,
+	}
+}
+
+func NewTestEnvironment() (*TestEnvironment, error) {
+	env := newTestEnvironment()
+
+	if err := env.start(); err != nil {
+		return nil, err
+	}
+
+	if err := env.initManager(); err != nil {
+		if stopErr := env.Stop(); err != nil {
+			return nil, kerrors.NewAggregate([]error{err, stopErr})
+		}
+		return nil, err
+	}
+
+	return env, nil
+}
+
+func (t *TestEnvironment) start() error {
+	cfg, err := t.Environment.Start()
+	t.Config = cfg
+
+	return err
+}
+
+func (t *TestEnvironment) initManager() error {
```

Comment:
IMO this should return the manager and the client

Reply:
I inlined this since we don't really need it as a separate method at all